### ISSUE TITLE
16$: Pretty error message for await-outside-async-function

### ISSUE THREAD

**Issue by strager** [quick-lint/quick-lint-js#140]
Outside of `async` functions, `await` is the name of a variable. However, if the programmer never declared a variable called `await`, and they used `await` in contexts where it would parse as a unary prefix operator, they probably intended to make the enclosing function an `async` function. In these cases, instead of reporting a reference-to-undeclared-variable error, report a use-of-await-outside-async-function error.

`yield` has similar rules with generator functions, except `yield` can be a unary prefix operator or a primary expression all on its own. We should report a use-of-yield-outside-generator-function error when appropriate for `yield`.

**Comment by strager** [comment]
It's hard for the parser to know whether a variable is declared or not. So this logic probably belongs in the linter.

However, the linter doesn't know how a variable is used. It can't really distinguish between know if `await*x` (where `await` can't be an operator) and `await+x` (where `await` can be an operator).

In the parser, if we see `await` followed by an identifier (or a few other things), that's definitely a syntax error within a non-async function. In this case, we can emit the use-of-await-outside-async-function error right from the parser.

---

In other words, there are a few cases to consider:

```
  inside async   |                            |
  or generator?  |  syntax                    |  error to report
-----------------+----------------------------+--------------------------------
  no             |  await  // declared        |  (none)
  no             |  await  // undeclared      |  await outside async func
  no             |  await + x  // declared    |  (none)
  no             |  await + x  // undeclared  |  await outside async func, or
                 |                            |  use of undeclared var 'await'
  no             |  await expr                |  await outside async func **
  async          |  await                     |  expected expression **
  async          |  await expr                |  (none)
  no             |  yield  // declared        |  (none)
  no             |  yield  // undeclared      |  yield outside generator func
  no             |  yield + x  // declared    |  (none)
  no             |  yield + x  // undeclared  |  yield outside async func, or
                 |                            |  use of undeclared var 'yield'
  no             |  yield expr                |  yield outside generator func **
  generator      |  yield                     |  (none)
  generator      |  yield expr                |  (none)
```

Errors marked with `**` I think can be detected in the parser alone. Errors not marked with `**` I think can only be detected in the linter.

**Comment by strager** [comment]
Implemented the parser-only solution in commit 28debd9d09d648eb8ca3afea4e56ff31f448539f. I suspect 99% of mistakes are detected by this. We should do some analysis on real code.

**Comment by strager** [comment]
I don't think the precise solution is worth implementing.
